src/ngMessageFormat/messageformat.js

```diff
@@ -0,0 +1,938 @@
+    this.rule = null;
+    return;
+  }
+  var token = match[0];
+  if (token[0] == "\\") {
+    // unescape next character and continue
+    this.interpolationParts.addText(this.text.substring(currentIndex, match.index) + token[1]);
+    return;
+  }
+  this.interpolationParts.addText(this.text.substring(currentIndex, match.index));
+  this.pushState();
+  this.ruleStack.push(this.ruleInterpolationEndMustache);
+  this.rule = this.ruleEnteredMustache;
+};
+
+MessageFormatParser.prototype.ruleInterpolationEndMustache = function ruleInterpolationEndMustache() {
+  var expressionFn = this.parsedFn;
+  this.popState();
+  this.interpolationParts.addExpressionFn(expressionFn);
+  this.rule = this.ruleInInterpolation;
+};
+
+var MUSTACHE_BEGIN_RE = /{{/g;
+MessageFormatParser.prototype.ruleMustache = function ruleMustache() {
+  var match = this.matchRe(MUSTACHE_BEGIN_RE);
+  if (match == null) {
+    var position = indexToLineAndColumn(this.text, this.index);
+    throw $$messageFormatMinErr('reqinterpbegin',
+        'Expecting {{ to start an interpolation at line {0}, column {1} of text "{2}"',
+        position.line, position.column, this.text);
+  }
+  this.rule = this.ruleEnteredMustache;
+};
+
+MessageFormatParser.prototype.ruleEnteredMustache = function ruleEnteredMustache() {
+  this.parsedFn = null;
+  this.ruleStack.push(this.ruleEndMustache);
+  this.rule = this.ruleAngularExpression;
+};
+
+MessageFormatParser.prototype.ruleEndMustacheInInterpolationOrMessage = function ruleEndMustacheInInterpolationOrMessage() {
+  var expressionFn = this.parsedFn;
+  this.popState();
+  this.interpolationParts.addExpressionFn(expressionFn);
+  this.rule = this.ruleInInterpolationOrMessageText;
+};
+
+
+
+var INTERP_END_RE = /\s*}}/g;
+MessageFormatParser.prototype.ruleEndMustache = function ruleEndMustache() {
+  var match = this.matchRe(INTERP_END_RE);
+  if (match == null) {
+    var position = indexToLineAndColumn(this.text, this.index);
+    throw $$messageFormatMinErr('reqendinterp',
+        'Expecting end of interpolation symbol, "{0}", at line {1}, column {2} in text "{3}"',
+        '}}', position.line, position.column, this.text);
+  }
+  if (this.parsedFn == null) {
+    this.parsedFn = this.$parse(this.expressionFn, this.stringifier);
+    if (goog.DEBUG) {
+      // Only needed in order to pretend to be $interpolate for tests copied from interpolateSpec.js
+      this.parsedFn.exp = this.expressionFn.exp;
+      this.parsedFn.expressions = this.expressionFn.expressions;
+    }
+  }
+  this.rule = null;
+};
+
+MessageFormatParser.prototype.ruleAngularExpression = function ruleAngularExpression() {
+  this.angularOperatorStack = [];
+  this.expressionStartIndex = this.index;
+  this.rule = this.ruleInAngularExpression;
+};
+
+function getEndOperator(opBegin) {
+  switch (opBegin) {
+    case "{": return "}";
+    case "[": return "]";
+    case "(": return ")";
+    default: return null;
+  }
+}
```

Comment:
A jsperf I wrote a little while back strongly favors the switch statement.  (<https://jsperf.com/map-vs-switch-vs-if>)  I don't particularly think one reads any clearer than the other so I went with the switch statement.